File: sdk-python/alure_sdk/receipt.py
from __future__ import annotations

import base64
import dataclasses
import hashlib
import json
from datetime import datetime, timezone, timedelta
from typing import Any

from .errors import ReceiptError


def _b64url_decode(value: str) -> bytes:
    padded = value.replace("-", "+").replace("_", "/")
    padded += "=" * (-len(padded) % 4)
    return base64.b64decode(padded)


@dataclasses.dataclass
class ReceiptValidationResult:
    valid: bool
    reason: str | None = None
    expires_at: str | None = None
    grace_period_days: int | None = None


class ReceiptVerifier:
    def __init__(self, public_key_pem: str | None = None):
        self.public_key_pem = public_key_pem

    def parse(self, token: str) -> dict[str, Any]:
        parts = token.split(".")
        if len(parts) != 3 or parts[0] != "v1":
            raise ReceiptError("invalid_receipt_format")
        try:
            payload_json = _b64url_decode(parts[1]).decode("utf-8")
            return json.loads(payload_json)
        except Exception as exc:
            raise ReceiptError("invalid_receipt_payload") from exc

    def verify_signature(self, token: str) -> bool:
        if not self.public_key_pem:
            raise ReceiptError("public_key_required")
        try:
            from cryptography.hazmat.primitives import serialization
            from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        except Exception as exc:
            raise ReceiptError("cryptography_missing") from exc

        parts = token.split(".")
        if len(parts) != 3 or parts[0] != "v1":
            return False
        payload_b64 = parts[1].encode("utf-8")
        signature = _b64url_decode(parts[2])
        public_key = serialization.load_pem_public_key(self.public_key_pem.encode("utf-8"))
        if not isinstance(public_key, Ed25519PublicKey):
            raise ReceiptError("invalid_public_key")
        try:
            public_key.verify(signature, payload_b64)
            return True
        except Exception:
            return False

    def validate_offline(
        self,
        token: str,
        device_id: str,
        now: datetime | None = None,
        verify_signature: bool = True,
    ) -> ReceiptValidationResult:
        payload = self.parse(token)
        if verify_signature:
            if not self.verify_signature(token):
                return ReceiptValidationResult(valid=False, reason="invalid_signature")

        device_hash = hashlib.sha256(device_id.encode("utf-8")).hexdigest()
        if payload.get("device_id_hash") != device_hash:
            return ReceiptValidationResult(valid=False, reason="device_mismatch")

        expires_at = payload.get("expires_at")
        grace_days = int(payload.get("grace_period_days") or 0)
        now_dt = now or datetime.now(timezone.utc)

        if expires_at:
            exp_dt = datetime.fromisoformat(expires_at.replace("Z", "+00:00"))
            if now_dt > exp_dt:
                grace_limit = exp_dt + timedelta(days=grace_days)
                if now_dt > grace_limit:
                    return ReceiptValidationResult(valid=False, reason="expired", expires_at=expires_at)
                return ReceiptValidationResult(
                    valid=True,
                    reason="grace_period",
                    expires_at=expires_at,
                    grace_period_days=grace_days,
                )

        return ReceiptValidationResult(valid=True, expires_at=expires_at, grace_period_days=grace_days)

File: sdk-rust/src/receipt.rs
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use chrono::{DateTime, Duration, Utc};
use ed25519_dalek::{Signature, VerifyingKey};
use pkcs8::DecodePublicKey;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::errors::ReceiptError;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReceiptValidationResult {
    pub valid: bool,
    pub reason: Option<String>,
    pub expires_at: Option<String>,
    pub grace_period_days: Option<i64>,
}

#[derive(Debug, Clone)]
pub struct ReceiptVerifier {
    public_key_pem: Option<String>,
}

impl ReceiptVerifier {
    pub fn new(public_key_pem: Option<String>) -> Self {
        Self { public_key_pem }
    }

    pub fn parse(&self, token: &str) -> Result<serde_json::Value, ReceiptError> {
        let parts: Vec<&str> = token.split('.').collect();
        if parts.len() != 3 || parts[0] != "v1" {
            return Err(ReceiptError("invalid_receipt_format".to_string()));
        }
        let payload_bytes = URL_SAFE_NO_PAD
            .decode(parts[1])
            .map_err(|_| ReceiptError("invalid_receipt_payload".to_string()))?;
        let payload = serde_json::from_slice(&payload_bytes)
            .map_err(|_| ReceiptError("invalid_receipt_payload".to_string()))?;
        Ok(payload)
    }

    pub fn verify_signature(&self, token: &str) -> Result<bool, ReceiptError> {
        let public_key_pem = self
            .public_key_pem
            .as_ref()
            .ok_or_else(|| ReceiptError("public_key_required".to_string()))?;
        let parts: Vec<&str> = token.split('.').collect();
        if parts.len() != 3 || parts[0] != "v1" {
            return Ok(false);
        }
        let payload = parts[1].as_bytes();
        let signature_bytes = URL_SAFE_NO_PAD
            .decode(parts[2])
            .map_err(|_| ReceiptError("invalid_signature".to_string()))?;
        let signature = Signature::from_slice(&signature_bytes)
            .map_err(|_| ReceiptError("invalid_signature".to_string()))?;
        let verifying_key = VerifyingKey::from_public_key_pem(public_key_pem)
            .map_err(|_| ReceiptError("invalid_public_key".to_string()))?;
        Ok(verifying_key.verify(payload, &signature).is_ok())
    }

    pub fn validate_offline(
        &self,
        token: &str,
        device_id: &str,
        now: Option<DateTime<Utc>>,
        verify_signature: bool,
    ) -> ReceiptValidationResult {
        let payload = match self.parse(token) {
            Ok(payload) => payload,
            Err(err) => {
                return ReceiptValidationResult {
                    valid: false,
                    reason: Some(err.0),
                    expires_at: None,
                    grace_period_days: None,
                }
            }
        };
        if verify_signature {
            match self.verify_signature(token) {
                Ok(true) => {}
                Ok(false) => {
                    return ReceiptValidationResult {
                        valid: false,
                        reason: Some("invalid_signature".to_string()),
                        expires_at: None,
                        grace_period_days: None,
                    }
                }
                Err(err) => {
                    return ReceiptValidationResult {
                        valid: false,
                        reason: Some(err.0),
                        expires_at: None,
                        grace_period_days: None,
                    }
                }
            }
        }

        let device_hash = Sha256::digest(device_id.as_bytes());
        let device_hash_hex = format!("{:x}", device_hash);
        if payload
            .get("device_id_hash")
            .and_then(|value| value.as_str())
            != Some(device_hash_hex.as_str())
        {
            return ReceiptValidationResult {
                valid: false,
                reason: Some("device_mismatch".to_string()),
                expires_at: None,
                grace_period_days: None,
            };
        }

        let expires_at = payload
            .get("expires_at")
            .and_then(|value| value.as_str())
            .map(str::to_string);
        let grace_days = payload
            .get("grace_period_days")
            .and_then(|value| value.as_i64())
            .unwrap_or(0);
        let now_dt = now.unwrap_or_else(Utc::now);
        if let Some(expires_at_str) = expires_at.clone() {
            if let Ok(exp_dt) = DateTime::parse_from_rfc3339(&expires_at_str) {
                let exp_dt = exp_dt.with_timezone(&Utc);
                if now_dt > exp_dt {
                    let grace_limit = exp_dt + Duration::days(grace_days);
                    if now_dt > grace_limit {
                        return ReceiptValidationResult {
                            valid: false,
                            reason: Some("expired".to_string()),
                            expires_at: Some(expires_at_str),
                            grace_period_days: Some(grace_days),
                        };
                    }
                    return ReceiptValidationResult {
                        valid: true,
                        reason: Some("grace_period".to_string()),
                        expires_at: Some(expires_at_str),
                        grace_period_days: Some(grace_days),
                    };
                }
            }
        }

        ReceiptValidationResult {
            valid: true,
            reason: None,
            expires_at,
            grace_period_days: Some(grace_days),
        }
    }
}
